---
title: 'mwsPLS: Multi-view weighted sparse PLS for miRNA-lncRNA-mRNA data analysis'
author: "Wenwen Min"
date: "2021/3/5"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load some R packages
```{r}
library("ggplot2")
library("cowplot")
library("ComplexHeatmap")
library("circlize")
print(getwd())
```

## Load BRCA miRNA+lncRNA+mRNA expression data
```{r}
load("../data/BRCA_miRNA_lncRNA_mRNA_Final.RData")
# ------------------------------------------------------------------------------
# BRCA = list()
# BRCA$mRNA_log_exp   = input_data$X1
# BRCA$miRNA_log_exp  = input_data$X2 
# BRCA$lncRNA_log_exp = input_data$X3
# ------------------------------------------------------------------------------
```
## Data preprocessing by scaling the columns of each numeric matrix
```{r}
# miRNA_log_exp: 751*581
X1 = scale(BRCA$miRNA_log_exp,  center = TRUE, scale = TRUE)
# lncRNA_log_exp: 751*3782
X2 = scale(BRCA$lncRNA_log_exp, center = TRUE, scale = TRUE)
# mRNA_log_exp: 751*6200
X3 = scale(BRCA$mRNA_log_exp,   center = TRUE, scale = TRUE)
```


## Import two mwsPLS functions
```{r}
# Multi_wsPLS_scheme_1_sum
Multi_wsPLS_scheme_1 = function(X1, X2, X3, k1, k2, k3, kn, Lc=0.1, niter=100, err=10^(-3), nstart=5){
  # ----------------------------------------------------------------------------
  # Input data
  # X \in R^{n \times p1} (n:samples, p1:variables)
  # Y \in R^{n \times p2} (n:samples, p2:variables)
  # Z \in R^{n \times p3} (n:samples, p3:variables)
  # Lc: Lipschitz constant
  # Sparse level: {ku, kv, kw, kn}
  # ----------------------------------------------------------------------------
  # The optimization problems solved are as follows
  # \sum_{i<j} \bm{u}_i^T\bm{X}_i^T\mbox{diag}(\bm{w})\bm{X}_j\bm{u}_j
  # Reference: Extensions of Sparse Canonical Correlation Analysis with Applications to Genomic Data
  # ----------------------------------------------------------------------------
  
  # Initialize optimal objective function
  opt_obj = 0
  
  # Initialize the objective function value of each initial point
  obj_nstart = rep(0,nstart); names(obj_nstart) = paste("init",1:nstart,sep = "_")
  
  for(init in 1:nstart){
    # Initialize u v and w
    set.seed(init)
    u1 = u1_0 = matrix(rnorm(ncol(X1)),ncol=1); u1 = u1/norm(u1,'E')
    u2 = u2_0 = matrix(rnorm(ncol(X2)),ncol=1); u2 = u2/norm(u2,'E')
    u3 = u3_0 = matrix(rnorm(ncol(X3)),ncol=1); u3 = u3/norm(u3,'E')
    w  = w0   = matrix(rep(1,nrow(X1)),ncol=1)  #Each element is one
    
    objs = c()
    # Iterative algorithm
    for (i in 1:niter){
      # update u1
      z1 = crossprod(X1, w*(X2%*%u2) + w*(X3%*%u3))
      u1 = PALM_sproject(z1, u1_0, Lc, k1)
      
      # update u2
      z2 = crossprod(X2, w*(X1%*%u1) + w*(X3%*%u3))
      u2 = PALM_sproject(z2, u2_0, Lc, k2)
      
      # update u3
      z3 = crossprod(X3, w*(X1%*%u1) + w*(X2%*%u2))
      u3 = PALM_sproject(z3, u3_0, Lc, k3)
      
      # update w
      z4 = (X1%*%u1)*(X2%*%u2) + (X1%*%u1)*(X3%*%u3) + (X2%*%u2)*(X3%*%u3)
      w  = PALM_update_w_l0_equal_k(z4, w0, Lc, kn)
      
      # Record objective function value
      # obj = t(w)%*%(((X1%*%u1)*(X2%*%u2)) + ((X1%*%u1)*(X2%*%u2)) + ((X1%*%u1)*(X2%*%u2))) 
      obj = t(w)%*%z4
      objs = c(objs,obj)
      
      # Algorithm termination condition
      if(i>=4&&abs(objs[i]-objs[i-1])<= err){
        break}else{
          u1_0=u1;u2_0=u2;u3_0=u3;w_0=w}
    }
    
    # --------------------------------------------------------------------------
    # Record objective every initialization
    obj_nstart[init] = obj
    # Keep optimal solution
    if(obj>opt_obj){
      opt_obj = obj
      opt_out = list(u1      = u1, 
                     u2      = u2,
                     u3      = u3,
                     w       = w, 
                     objs    = objs, 
                     opt_obj = opt_obj)
      
    }
  }
  
  # ----------------------------------------------------------------------------
  # Output
  out = list(u1         = opt_out$u1, 
             u2         = opt_out$u2, 
             u3         = opt_out$u3, 
             w          = opt_out$w, 
             objs       = opt_out$objs, 
             opt_obj    = opt_out$opt_obj,
             obj_nstart = obj_nstart)
  return(out)
}

# ------------------------------------------------------------------------------
# Sparse project function
# s.t. ||w||_0 <= k_w,  0<=w_i<=1
PALM_sproject = function(z, u0, Lc, k){
  #Lc is the Lipschitz constant
  z = u0 + z/Lc
  if((length(z)<k)|(sum(z^2)==0)) return(z)  
  u = abs(z);
  u[-order(u,decreasing=T)[1:k]] = 0
  u = u/sqrt(sum(u^2))
  u = sign(z)*u
  return(u) 
}
# ------------------------------------------------------------------------------
# Learning w
# s.t. ||w||_0 <= k_w,  w_i \in {0,1}^n
PALM_update_w_l0_equal_k = function(z, w0, Lc, k){
  #Lc is the Lipschitz constant
  z = w0 + z/Lc
  if((length(z)<=k)|(sum(z^2)==0)) {
    w = matrix(rep(1,length(z)),ncol=1)
    return(w) }
  
  w = matrix(rep(1,length(z)),ncol=1) # ones
  w[-order(z,decreasing=T)[1:k]] = 0
  
  return(w)
}
# ------------------------------------------------------------------------------

# Multi_wsPLS_scheme_2_multiply
Multi_wsPLS_scheme_2 = function(X1, X2, X3, k1, k2, k3, kn, Lc=0.1, niter=100, err=10^(-3), nstart=5){
  # ----------------------------------------------------------------------------
  # Input data
  # X \in R^{n \times p1} (n:samples, p1:variables)
  # Y \in R^{n \times p2} (n:samples, p2:variables)
  # Z \in R^{n \times p3} (n:samples, p3:variables)
  # Lc: Lipschitz constant
  # Sparse level: {ku, kv, kw, kn}
  # ----------------------------------------------------------------------------
  # The optimization problems solved are as follows
  # \begin{equation}\label{equ-15}
  # \begin{aligned}
  # & \underset{\bm{u}_i, \bm{w}}{\text{minimize}} && -\bm{w}^{\rm{T}}\big[\bigodot\limits_{i=1}^M(\bm{X}_i\bm{u}_i)\big] \\
  # & \text{subject to} &&\|\bm{u}_i\| = 1,\|\bm{u}_i\|_0\leq k_i~\forall  i \\
  # &                   &&\|\bm{w}\|_0\leq k_w, \|\bm{w}\|_\infty \leq 1, \bm{w}\geq 0,\\
  # \end{aligned}
  # \end{equation}
  # ----------------------------------------------------------------------------
  
  # Initialize optimal objective function
  opt_obj = 0
  
  # Initialize the objective function value of each initial point
  obj_nstart = rep(0,nstart); names(obj_nstart) = paste("init",1:nstart,sep = "_")
  
  for(init in 1:nstart){
    # Initialize u v and w
    set.seed(init)
    u1 = u1_0 = matrix(rnorm(ncol(X1)),ncol=1); u1 = u1/norm(u1,'E')
    u2 = u2_0 = matrix(rnorm(ncol(X2)),ncol=1); u2 = u2/norm(u2,'E')
    u3 = u3_0 = matrix(rnorm(ncol(X3)),ncol=1); u3 = u3/norm(u3,'E')
    w  = w0   = matrix(rep(1,nrow(X1)),ncol=1)  #Each element is one
    
    objs = c()
    # Iterative algorithm
    for (i in 1:niter){
      # update u1
      z1 = crossprod(X1, w*(X2%*%u2)*(X3%*%u3))
      u1 = PALM_sproject(z1, u1_0, Lc, k1)
      
      # update u2
      z2 = crossprod(X2, w*(X1%*%u1)*(X3%*%u3))
      u2 = PALM_sproject(z2, u2_0, Lc, k2)
      
      # update u3
      z3 = crossprod(X3, w*(X1%*%u1)*(X2%*%u2))
      u3 = PALM_sproject(z3, u3_0, Lc, k3)
      
      # update w
      z4 = (X1%*%u1)*(X2%*%u2)*(X3%*%u3)
      w  = PALM_update_w_l0_equal_k(z4, w0, Lc, kn)
      
      # Record objective function value
      obj = sum((X1%*%u1)*(X2%*%u2)*(X3%*%u3)*w)
      objs = c(objs,obj)
      
      # Algorithm termination condition
      if(i>=4&&abs(objs[i]-objs[i-1])<= err){
        break}else{
          u1_0=u1;u2_0=u2;u3_0=u3;w_0=w}
    }
    
    # --------------------------------------------------------------------------
    # Record objective every initialization
    obj_nstart[init] = obj
    # Keep optimal solution
    if(obj>opt_obj){
      opt_obj = obj
      opt_out = list(u1      = u1, 
                     u2      = u2,
                     u3      = u3,
                     w       = w, 
                     objs    = objs, 
                     opt_obj = opt_obj)
      
    }
  }
  
  # ----------------------------------------------------------------------------
  # Output
  out = list(u1         = opt_out$u1, 
             u2         = opt_out$u2, 
             u3         = opt_out$u3, 
             w          = opt_out$w, 
             objs       = opt_out$objs, 
             opt_obj    = opt_out$opt_obj,
             obj_nstart = obj_nstart)
  return(out)
}
```

## Display data dimensions
```{r}
# 
Data_dim = c(
  number.of.sample = dim(X1)[1],
  number.of.miRNA  = dim(X1)[2],
  number.of.lncRNA = dim(X2)[2],
  number.of.mRNA   = dim(X3)[2])
Data_dim
```

## Application mwsPLS to BRCA miRNA-lncRNA-mRNA expression data
```{r}

# Setting parameters of mwsPLS
k1 = 30   # Number of miRNAs
k2 = 100  # Number of lncRNAs
k3 = 200  # Number of mRNAs
kn = 190  # Number of samples

# Multi_wsPLS_scheme_1_sum
out1 = Multi_wsPLS_scheme_1(X1, X2, X3, k1, k2, k3, kn, Lc=1, niter=20, err=10^(-3), nstart=5)

# Multi_wsPLS_scheme_2_multiply
out2 = Multi_wsPLS_scheme_2(X1, X2, X3, k1, k2, k3, kn, Lc=1, niter=20, err=10^(-3), nstart=5)
```


## Result analysis
### Show the convergence of the objective function of mwsPLS
```{r, fig.width=10, fig.height=5, fig.fullwidth=TRUE}
plot_obj = function(out4){
  obj_df = data.frame(Iterations=1:length(out4$objs), obj=-out4$objs)
  
  fig_obj = ggplot(obj_df, aes(x=Iterations, y=obj)) + 
    geom_line() + geom_point(shape=1) +
    xlab("Iterations") + ggtitle(NULL) + ylab("Obj") # + theme_classic()
  
  fig_obj = fig_obj + theme(
    axis.title.y = element_text(size = rel(1.5)),
    axis.title.x = element_text(size = rel(1.5)),
    axis.text=element_text(size = rel(1),colour="black"),
    axis.line  = element_line(size = rel(1),colour = "black")) 
  fig_obj
}

# An R package is used to merge two ggplot plots
library(gridExtra)

# mwsPLS: The objective function is designed in the form of sum
fig1 = plot_obj(out1) + ggtitle("mwsPLS_sum")
ggsave("../result/fig_obj_mwsPLS_sum.png", width = 5, height = 5)

# mwsPLS: The objective function is designed in the form of multiply
fig2 = plot_obj(out2) + ggtitle("mwsPLS_multiply")
ggsave("../result/fig_obj_mwsPLS_multiply.png", width = 5, height = 5)

# merge two ggplot plots
grid.arrange(fig1, fig2, ncol=2)

# ggsave("Fig_obj.png", width = 5, height = 5)
```

### Projection correlation calculation
```{r}
get_cors = function(out){
  cors = c(gene_miR    = cor((X1%*%out$u1)[which(out$w!=0)],(X2%*%out$u2)[which(out$w!=0)]),
           gene_lncRNA = cor((X1%*%out$u1)[which(out$w!=0)],(X3%*%out$u3)[which(out$w!=0)]),
           miR_lncRNA  = cor((X2%*%out$u2)[which(out$w!=0)],(X3%*%out$u3)[which(out$w!=0)]))
  return(cors) # print(cors)
}

print("mwsPLS: The objective function is designed in the form of sum")
cor_mwsPLS_sum = get_cors(out1)
cor_mwsPLS_sum

print("mwsPLS: The objective function is designed in the form of multiply")
cors_mwsPLS_multiply = get_cors(out2)
cors_mwsPLS_multiply
```
### Obtain members of miRNA-lncRNA-mRNA co-module
```{r}
# Obtain members of the first miRNA-lncRNA-mRNA co-module identified by mwsPLS_sum
# out1 = Multi_wsPLS_scheme_1(X1, X2, X3, k1, k2, k3, kn, Lc=1, niter=20, err=10^(-3), nstart=5)
obtain_members_of_comodule = function(X1, X2, X3, out1){
  # We first obtain three-dimensional biomolecule names
  miRNA_name  = colnames(X1)
  lncRNA_name = colnames(X2) 
  mRNA_name   = colnames(X3) 
  sample_name = rownames(X1)
  
  
  comodule=list()
  comodule$miRNA  = miRNA_name [which(out1$u1!=0)] 
  comodule$lncRNA = lncRNA_name[which(out1$u2!=0)]
  comodule$mRNA   = mRNA_name  [which(out1$u3!=0)]
  comodule$sample = sample_name[which(out1$ w!=0)]
  
  return(comodule)
}

coModule1_mwsPLS_sum = obtain_members_of_comodule(X1, X2, X3, out1)
coModule1_mwsPLS_multiply = obtain_members_of_comodule(X1, X2, X3, out2)
```

### The average absolute value of Pearson correlation coefficients for all miRNA-lncRNA, miRNA-mRNA, and lncRNA-mRNA pairs
```{r}
compute_S_score = function(X1, X2, X3, coModule1_mwsPLS_sum){
  
  # Obtain three sub-matrices 
  X1_sub = X1[coModule1_mwsPLS_sum$sample, coModule1_mwsPLS_sum$miRNA]
  X2_sub = X2[coModule1_mwsPLS_sum$sample, coModule1_mwsPLS_sum$lncRNA]
  X3_sub = X3[coModule1_mwsPLS_sum$sample, coModule1_mwsPLS_sum$mRNA]
  
  # Pearson correlation coefficients for all miRNA-lncRNA, miRNA-mRNA, and lncRNA-mRNA pairs
  miRNA_lncRNA_corMat = cor(X1_sub, X2_sub)
  miRNA_gene_corMat   = cor(X1_sub, X3_sub)
  lncRNA_gene_corMat  = cor(X2_sub, X3_sub)
  
  # The average absolute value of Pearson correlation coefficients
  all_pair = c(miRNA_lncRNA_corMat,miRNA_gene_corMat,lncRNA_gene_corMat)
  S0 = mean(abs(all_pair))
  return(S0)
}

S0_coModule1_mwsPLS_sum = compute_S_score(X1, X2, X3, coModule1_mwsPLS_sum)
S0_coModule1_mwsPLS_sum
S0_coModule1_mwsPLS_multiply = compute_S_score(X1, X2, X3, coModule1_mwsPLS_multiply)
S0_coModule1_mwsPLS_multiply
```
### Permutation test 
To investigate the statistical significance, we adopt a permutation test by shuffling these miRNAs, lncRNAs, mRNAs and samples according to those identified modules, and then compute the average correlation score for them. 
(Permutation test: https://bookdown.org/curleyjp0/psy317l_guides/permutation-testing.html)
```{r}
Permutation_test_coModule = function(X1, X2, X3, k_mir, k_lnc, k_mrna, k_sample, n_try = 1000){
  
  # initialization
  scores = c()
  
  # number of permutation tests
  for(i in 1:n_try){
    
    # Get the dimensions of the three matrices
    n  = dim(X1)[1]
    p1 = dim(X1)[2]
    p2 = dim(X2)[2]
    p3 = dim(X3)[2]
    
    # ids from a random sampling module  
    set.seed(i)
    miRs    = sample(1:p1, k_mir,    replace =F)
    lncRNAs = sample(1:p2, k_lnc,    replace =F)
    mRNAs   = sample(1:p3, k_mrna,   replace =F)
    samples = sample(1:n,  k_sample, replace =F)
    
    # obtain three sub-matrices 
    X1_sub = X1[samples, miRs]
    X2_sub = X2[samples, lncRNAs]
    X3_sub = X3[samples, mRNAs]
    
    # pearson correlation coefficients for all miRNA-lncRNA, miRNA-mRNA, and lncRNA-mRNA pairs
    miRNA_lncRNA_corMat = cor(X1_sub, X2_sub)
    miRNA_gene_corMat   = cor(X1_sub, X3_sub)
    lncRNA_gene_corMat  = cor(X2_sub, X3_sub)
    
    # the average absolute value of Pearson correlation coefficients
    all_pair = c(miRNA_lncRNA_corMat,miRNA_gene_corMat,lncRNA_gene_corMat)
    score = mean(abs(all_pair))
    
    # insert
    scores = c(scores,score)
  }
  return(scores)
}

# get average absolute values of 1000 random co-modules
scores = Permutation_test_coModule(X1, X2, X3, k_mir=k1, k_lnc=k2, k_mrna=k3, k_sample=kn, n_try = 1000)

# compute p-value
p_val_coModule1_mwsPLS_sum = (sum(scores > S0_coModule1_mwsPLS_sum) + 1)/1001
p_val_coModule1_mwsPLS_sum

p_val_coModule1_multiply = (sum(scores > S0_coModule1_mwsPLS_multiply) + 1)/1001
p_val_coModule1_multiply
```

## plot histogram 
```{r}
# plot histogram using ggplot2
dat  = data.frame(x=1:length(scores),y=scores)
fig1 = ggplot(data=dat, aes(y)) +
  # plese note: set bins = 1000
  geom_histogram(fill = "steelblue", bins = 1000, colour = "navy", alpha = 0.9) + 
  labs(x="Average absolute value of PCCs", y="Count", title="Permutation test")

fig2 = fig1 + geom_vline(xintercept = S0_coModule1_mwsPLS_sum, colour = "red", lwd=1, lty=2) +
  geom_vline(xintercept = S0_coModule1_mwsPLS_multiply, colour = "blue", lwd=1, lty=2)

fig2

ggsave("../result/fig_histogram_permutation_test.png", width = 8, height = 5, units = "in")
```

### Plot miRNA-lncRNA-mRNA co-module heatmap
load co_module_heatmap function
```{r}
co_module_heatmap = function(X1, X2, X3, out, figName=NULL){
  # Input:
  # X1 = scale(BRCA$miRNA_log_exp,  center = TRUE, scale = TRUE)
  # X2 = scale(BRCA$lncRNA_log_exp, center = TRUE, scale = TRUE)
  # X3 = scale(BRCA$mRNA_log_exp,   center = TRUE, scale = TRUE)
  # out is the output of Multi_wsPLS_scheme_1
  
  # ----------------------------------------------------------------------------
  # compute number of nonzero elements
  kn = length(which(out$w !=0))
  k1 = length(which(out$u1!=0))
  k2 = length(which(out$u2!=0))
  k3 = length(which(out$u3!=0))
  
  # ----------------------------------------------------------------------------
  # coModule + random features 
  fold_num = 3
  n_max  = min(length(out$w ), kn*fold_num) 
  p1_max = min(length(out$u1), k1*fold_num) 
  p2_max = min(length(out$u2), k2*fold_num) 
  p3_max = min(length(out$u3), k3*fold_num) 
  
  w_order  = order(-abs(out$w ))[1:n_max]
  u1_order = order(-abs(out$u1))[1:500]
  u2_order = order(-abs(out$u2))[1:500]
  u3_order = order(-abs(out$u3))[1:500]
  
  # ----------------------------------------------------------------------------
  # Partially_shuffle_order
  Partially_shuffle_order = function(vec, k, seed0=1){
    vec1 = vec[1:k]
    vec2 = vec[-c(1:k)]
    set.seed(seed0)
    vec2 = sample(vec2, length(vec2), replace = FALSE)
    fin_vec = c(vec1,vec2)
  }
  
  w_order  = Partially_shuffle_order(w_order ,kn)
  u1_order = Partially_shuffle_order(u1_order,k1)
  u2_order = Partially_shuffle_order(u2_order,k2)
  u3_order = Partially_shuffle_order(u3_order,k3)
  
  # ----------------------------------------------------------------------------
  # obtain the sort matrix
  X1_order = X1[w_order, u1_order]
  X2_order = X2[w_order, u2_order]
  X3_order = X3[w_order, u3_order]
  
  n = dim(X1_order)[1]
  p = dim(X1_order)[2]
  q = dim(X2_order)[2]
  t = dim(X3_order)[2]
  
  # ----------------------------------------------------------------------------
  # Set the color parameters of the heat map
  col_fun = colorRamp2(c(-2, 0, 2), c("blue","white", "red"))
  col_fun(seq(0, 2))
  
  png(filename = figName, width=9.15*3/2, height=4.5, units = "in",res=600)
  ht1 = Heatmap(X1_order,
                cluster_rows = F,
                cluster_columns = F,
                #column_names_side = "top",
                #row_names_side = "left",
                show_row_names = F,
                show_column_names = F,
                row_title = "sample",
                column_title = "miRNA", col = col_fun,
                show_heatmap_legend = F, name="foo1")
  
  ht2 = Heatmap(X2_order,
                cluster_rows = F,
                cluster_columns = F,
                #column_names_side = "top",
                #row_names_side = "left",
                show_row_names = F,
                show_column_names = F,
                column_title = "lncRNA",col = col_fun,
                show_heatmap_legend = T, name="Exp")
  
  ht3 = Heatmap(X3_order,
                cluster_rows = F,
                cluster_columns = F,
                #column_names_side = "top",
                #row_names_side = "left",
                show_row_names = F,
                show_column_names = F,
                column_title = "mRNA",col = col_fun,
                show_heatmap_legend = F, name="foo3")
  
  
  ht_list = ht1 + ht2 + ht3
  ht = draw(ht_list, auto_adjust = T)
  
  decorate_heatmap_body("foo1", {
    grid.rect(gp = gpar(fill = "transparent", col = "black", lwd = 2))
    grid.rect(x = unit(0/p, "npc"), width = unit(k1/p, "npc"),
              y = unit((n-kn)/n, "npc"), height = unit(kn/n, "npc"),
              hjust = 0, vjust = 0,
              gp=gpar(fill = "transparent", col = "red", lwd = 2))
  })
  
  decorate_heatmap_body("Exp", {
    grid.rect(gp = gpar(fill = "transparent", col = "black", lwd = 2))
    grid.rect(x = unit(0/q, "npc"), width = unit(k2/q, "npc"),
              y = unit((n-kn)/n, "npc"), height = unit(kn/n, "npc"),
              hjust = 0, vjust = 0,
              gp=gpar(fill = "transparent", col = "red", lwd = 2))
  })
  
  decorate_heatmap_body("foo3", {
    grid.rect(gp = gpar(fill = "transparent", col = "black", lwd = 2))
    grid.rect(x = unit(0/t, "npc"), width = unit(k3/t, "npc"),
              y = unit((n-kn)/n, "npc"), height = unit(kn/n, "npc"),
              hjust = 0, vjust = 0,
              gp=gpar(fill = "transparent", col = "red", lwd = 2))
  })
  dev.off()
  
  return(ht)
}
```


#### Plot miRNA-lncRNA-mRNA co-module heatmap 
```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# Plot miRNA-lncRNA-mRNA co-module heatmap for mwsPLS_sum
ht_sum = co_module_heatmap(X1, X2, X3, out1, figName="../result/fig_heatmap_comodule1_mwsPLS_sum.png") 

# Plot miRNA-lncRNA-mRNA co-module heatmap for mwsPLS_multiply
ht_multiply = co_module_heatmap(X1, X2, X3, out2, figName="../result/fig_heatmap_comodule1_mwsPLS_multiply.png") 
```


### Plot the correlation matrix of any two types of molecules
```{r}
cor_mat_heatmap = function(X1, X2, X3, out, corHeatmapNames){
  # ----------------------------------------------------------------------------
  # Input:
  # X1 = scale(BRCA$miRNA_log_exp,  center = TRUE, scale = TRUE)
  # X2 = scale(BRCA$lncRNA_log_exp, center = TRUE, scale = TRUE)
  # X3 = scale(BRCA$mRNA_log_exp,   center = TRUE, scale = TRUE)
  # out is the output of Multi_wsPLS_scheme_1
  # ----------------------------------------------------------------------------
  
  # compute number of nonzero elements
  kn = length(which(out$w !=0))
  k1 = length(which(out$u1!=0))
  k2 = length(which(out$u2!=0))
  k3 = length(which(out$u3!=0))
  
  # coModule + random features 
  fold_num = 3
  n_max  = min(length(out$w ), kn*fold_num) 
  p1_max = min(length(out$u1), k1*fold_num) 
  p2_max = min(length(out$u2), k2*fold_num) 
  p3_max = min(length(out$u3), k3*fold_num) 
  
  w_order  = order(-abs(out$w ))[1:n_max]
  u1_order = order(-abs(out$u1))[1:p1_max]
  u2_order = order(-abs(out$u2))[1:p2_max]
  u3_order = order(-abs(out$u3))[1:p3_max]
  
  # Partially_shuffle_order
  Partially_shuffle_order = function(vec, k, seed0=1){
    vec1 = vec[1:k]
    vec2 = vec[-c(1:k)]
    set.seed(seed0)
    vec2 = sample(vec2, length(vec2), replace = FALSE)
    fin_vec = c(vec1,vec2)
  }
  
  w_order  = Partially_shuffle_order(w_order ,kn)
  u1_order = Partially_shuffle_order(u1_order,k1)
  u2_order = Partially_shuffle_order(u2_order,k2)
  u3_order = Partially_shuffle_order(u3_order,k3)
  
  # obtain the sort matrix
  X1_order = X1[w_order,u1_order]
  X2_order = X2[w_order,u2_order]
  X3_order = X3[w_order,u3_order]
  
  n = dim(X1_order)[1]
  p = dim(X1_order)[2]
  q = dim(X2_order)[2]
  t = dim(X3_order)[2]
  
  # three_cormat_heatmap
  miRNA_lncRNA_corMat = cor(X1_order[1:kn, ], X2_order[1:kn, ])
  miRNA_gene_corMat   = cor(X1_order[1:kn, ], X3_order[1:kn, ])
  lncRNA_gene_corMat  = cor(X2_order[1:kn, ], X3_order[1:kn, ])
  
  # plot heatmap
  plot_a_cormat_heatmap = function(corMat, p, q, k1, k2,
                                   fileName, rowName="miRNA",colName="Gene"){
    # abs
    corMat = abs(corMat)
    
    # Set the color parameters of the heat map
    # col_fun = colorRamp2(seq(min(corMat)/2, max(corMat)/2, length = 2), c("blue", "#EEEEEE", "red"), space = "RGB")
    
    png(filename = fileName, width=4.8, height=4.5, units = "in",res=600)
    ht = Heatmap(corMat, 
                 cluster_rows = F,
                 cluster_columns = F, 
                 show_row_names = F,
                 show_column_names = F,
                 row_title = rowName,
                 column_title = colName,
                 # col = col_fun,
                 show_heatmap_legend = T, name="cor")
    
    ht = draw(ht)
    decorate_heatmap_body("cor", {
      grid.rect(gp = gpar(fill = "transparent", col = "black", lwd = 2))
      grid.rect(x = unit(0/q, "npc"), width = unit(k2/q, "npc"),
                y = unit((p-k1)/p, "npc"), height = unit(k1/p, "npc"),
                hjust = 0, vjust = 0,
                gp=gpar(fill = "transparent", col = "red", lwd = 2))
    })
    dev.off()
    return(ht)
  }
  #-----------------------------------------------------------------------------
  
  corFig1 = plot_a_cormat_heatmap(miRNA_lncRNA_corMat, p, q, k1, k2,
                                  corHeatmapNames[1], rowName="miRNA", colName="lncRNA")
  
  corFig2 = plot_a_cormat_heatmap(miRNA_gene_corMat, p, t, k1, k3,
                                  corHeatmapNames[2], rowName="miRNA",colName="mRNA")
  
  corFig3 = plot_a_cormat_heatmap(lncRNA_gene_corMat, q, t, k2, k3,
                                  corHeatmapNames[3], rowName="lncRNA",colName="mRNA")
  
  
  return(ht_list=list(corFig1=corFig1, corFig2=corFig2, corFig3=corFig3))
}

```


#### Plot corMat heatmaps 
```{r, fig.width=5, fig.height=5, out.width="30%"}
#### Plot corMat heatmaps for mwsPLS_sum
corHeatmapNames1 = c(name1 = "../result/fig_heatmap_corMat_mwsPLS_sum_[miRNA-lncRNA].png",
                     name2 = "../result/fig_heatmap_corMat_mwsPLS_sum_[miRNA-mRNA].png",
                     name3 = "../result/fig_heatmap_corMat_mwsPLS_sum_[lncRNA-mRNA].png")

ht_corMat_sum = cor_mat_heatmap(X1, X2, X3, out1, corHeatmapNames1) 

# Plot corMat heatmaps for mwsPLS_multiply
corHeatmapNames2 = c(name1 = "../result/fig_heatmap_corMat_mwsPLS_multiply_[miRNA-lncRNA].png",
                     name2 = "../result/fig_heatmap_corMat_mwsPLS_multiply_[miRNA-mRNA].png",
                     name3 = "../result/fig_heatmap_corMat_mwsPLS_multiply_[lncRNA-mRNA].png")

ht_corMat_multiply = cor_mat_heatmap(X1, X2, X3, out2, corHeatmapNames2) 
```